### PR TITLE
Refactor examples

### PR DESCRIPTION
**Pull request by siuwuncheung**
- The DMD prediction timers are adjusted to only count the prediction in the final time. Previously, since timers count all the intermediate predictions for the purpose of visualizations, the DMD prediction time is significantly different if visualization is enabled or disabled.
- `BasisGeneratorFinalSummary` in two MFEM PROM examples is replaced by `finalSummary` in `BasisGenerator` class. (`BasisGeneratorFinalSummary` in Laghos can be replaced as well.)
-  In `BasisGenerator` class, `loadSampleRange` routine is added to enable the capability of using selected range of snapshot indices from `col_min` to `col_max` inclusive. The `loadSample` routine is now based on `loadSampleRange` using all indices.

### CODE REVIEW

**Comment by siuwuncheung** [comment]
@jtlau I'm using the general `combine_samples` script for one of my projects, and would like to ask for your PR to make sure all these changes look alright to you.

**Review by jtlau — APPROVED** [review]
`combine_samples` looks good to me. Thanks for updating it!

File: lib/linalg/BasisGenerator.cpp
Comment (siuwuncheung) [inline lib/linalg/BasisGenerator.cpp:328]: @axla-io I found there is a merge conflict in this PR, after your recently merged PR on nonlinear elasticity. It is due to you made some changes in the `BasisGeneratorFinalSummary` and I moved it to the `BasisGenerator` class. Is it safe to simply copy your changes here?

Comment (axla-io) [reply lib/linalg/BasisGenerator.cpp:328]: Not sure if I understand the question and I'm a bit busy with a deadline. Can we talk about this on Thursday?

Comment (siuwuncheung) [reply lib/linalg/BasisGenerator.cpp:328]: Sure. No rush. Let's talk about it later.

**Comment by chldkdtn** [comment]
This PR needs one more reviewer, @ckendrick @dylan-copeland @dreamer2368 , anyone volunteer?

**Review by ckendrick — APPROVED** [review]

**Comment by dylan-copeland** [comment]
> This PR needs one more reviewer, @ckendrick @dylan-copeland @dreamer2368 , anyone volunteer?

This is a WIP, not RFR.

**Comment by chldkdtn** [comment]
> > This PR needs one more reviewer, @ckendrick @dylan-copeland @dreamer2368 , anyone volunteer?
> 
> This is a WIP, not RFR.

Oh, I thought this was RFR. @siuwuncheung Can you confirm?

File: lib/linalg/BasisGenerator.h
Comment (dylan-copeland) [inline lib/linalg/BasisGenerator.h:26]: Can this be included in the cpp file instead?

File: lib/linalg/BasisGenerator.h
Comment (dylan-copeland) [inline lib/linalg/BasisGenerator.h:154]: Comment that the column indices are 1-based? Or should we change to 0-based? I suppose this is a rare case where 1-based is more natural, so we can keep it as 1-based.

File: lib/linalg/BasisGenerator.h
Comment (dylan-copeland) [inline lib/linalg/BasisGenerator.h:172]: `bases` -> `basis vectors`

File: lib/linalg/BasisGenerator.cpp
Comment (dylan-copeland) [inline lib/linalg/BasisGenerator.cpp:331]: Pass the string by reference.

File: lib/linalg/BasisGenerator.h
Comment (dylan-copeland) [inline lib/linalg/BasisGenerator.h:295]: Should we input `1 - energyFraction` instead, with an appropriate name? It would be easier to put 1e-6 than 0.999999.

File: examples/misc/combine_samples.cpp
Comment (dylan-copeland) [inline examples/misc/combine_samples.cpp:56]: Comment that column indices are 1-based.

File: lib/linalg/BasisGenerator.h
Comment (dylan-copeland) [inline lib/linalg/BasisGenerator.h:289]: `difference of energy fraction from 1.0` or maybe clearer is
`energy fraction = 1.0 - energyFractionThreshold`

**Pull request merged**
